src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+                let _ = sender.send(response_message);
+            } else {
+                // we seem to have received a duplicate of the response message,
+                // ignoring it
+            }
+        }
+    }
+
+    fn remove_pending_request(&self, reqid: &UUID) -> Option<Sender<UMessage>> {
+        self.pending_requests
+            .lock()
+            .map_or(None, |mut pending_requests| pending_requests.remove(reqid))
+    }
+
+    #[cfg(test)]
+    fn contains(&self, reqid: &UUID) -> bool {
+        self.pending_requests
+            .lock()
+            .map_or(false, |pending_requests| {
+                pending_requests.contains_key(reqid)
+            })
+    }
+}
+
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if msg.is_response() {
+            if let Some(reqid) = msg
+                .attributes
+                .as_ref()
+                .and_then(|attribs| attribs.reqid.clone().into_option())
+            {
+                self.handle_response(&reqid, msg);
+            }
+        }
+    }
+}
+
+/// An ['RpcClient'] which keeps all information about pending requests in memory.
+pub struct InMemoryRpcClient {
+    transport: Arc<dyn UTransport>,
+    uri_provider: Arc<dyn LocalUriProvider>,
+    response_listener: Arc<ResponseListener>,
+}
+
+impl InMemoryRpcClient {
+    /// Creates a new RPC client for a given transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The uProtocol Transport Layer implementation to use for invoking service operations.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the generic RPC Response listener could not be
+    /// registered with the given transport.
+    pub async fn new(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
+    ) -> Result<Self, UStatus> {
+        let response_listener = Arc::new(ResponseListener {
+            pending_requests: Mutex::new(HashMap::new()),
+        });
+        transport
+            .register_listener(
+                &UUri::any(),
+                Some(&uri_provider.get_source_uri()),
+                response_listener.clone(),
+            )
+            .await?;
+
+        Ok(InMemoryRpcClient {
+            transport,
+            uri_provider,
+            response_listener,
+        })
+    }
+
+    #[cfg(test)]
+    fn contains_pending_request(&self, reqid: &UUID) -> bool {
+        self.response_listener.contains(reqid)
+    }
+}
+
+#[async_trait]
+impl RpcClient for InMemoryRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
+        call_options: CallOptions,
+        payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        let message_id = call_options.message_id().unwrap_or_else(UUID::build);
```

Comment:
Might just be me, but I think it'd increase cohesion if we moved this line to right above where we call
```rust
builder.with_message_id(message_id.clone());
```

Reply:
FMPOV it is at the right position because the variable is also needed/used at the end of the function and hiding it within the creation of the builder makes it harder to spot.